File: orchestrator-bridge/src/main/kotlin/com/orchestrator/mcp/bridge/local/ServerProcess.kt
package com.orchestrator.mcp.bridge.local

import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import org.slf4j.LoggerFactory

/**
 * Lifecycle states for a local MCP server process.
 */
enum class ServerState {
    STARTING, READY, ACTIVE, CRASHED, RESTARTING, STOPPING, DEAD, FAILED
}

/**
 * Configuration for a single local MCP server.
 */
data class ServerConfig(
    val command: String,
    val args: List<String> = emptyList(),
    val env: Map<String, String> = emptyMap(),
    val timeout: Long = 30_000,
    val maxRetries: Int = 3,
    val disabled: Boolean = false,
)

/**
 * MCP tool definition from a local server.
 */
data class ToolDefinition(
    val name: String,
    val description: String? = null,
    val inputSchema: JsonObject? = null,
)

/**
 * Single local MCP server process with state machine lifecycle.
 * Manages spawn, initialize, health check, and restart for one server.
 */
class ServerProcess(
    val name: String,
    private val config: ServerConfig,
    private val scope: CoroutineScope,
) {
    private val logger = LoggerFactory.getLogger("local:$name")
    private val rpc = StdioJsonRpc()
    private var process: Process? = null
    private var state: ServerState = ServerState.STARTING
    private var retryCount = 0
    private var tools: List<ToolDefinition> = emptyList()
    var onCrashed: ((String) -> Unit)? = null

    val currentState: ServerState get() = state
    val currentTools: List<ToolDefinition> get() = tools

    suspend fun start(): Boolean {
        state = ServerState.STARTING
        return try {
            spawnProcess()
            if (!initialize()) {
                state = ServerState.FAILED
                return false
            }
            state = ServerState.READY
            if (fetchTools()) state = ServerState.ACTIVE
            true
        } catch (e: Exception) {
            logger.error("Start failed: {}", e.message)
            state = ServerState.FAILED
            false
        }
    }

    suspend fun stop() {
        state = ServerState.STOPPING
        killProcess()
        tools = emptyList()
    }

    suspend fun restart(): Boolean {
        if (retryCount >= config.maxRetries) {
            state = ServerState.DEAD
            logger.error("Max retries ({}) exceeded → DEAD", config.maxRetries)
            return false
        }
        state = ServerState.RESTARTING
        val delay = minOf(1000L * (1 shl retryCount), 30_000L)
        retryCount++
        logger.info("Restart #{} in {}ms", retryCount, delay)
        delay(delay)
        return start()
    }

    suspend fun callTool(toolName: String, args: JsonObject): JsonElement? {
        val params = buildJsonObject {
            put("name", JsonPrimitive(toolName))
            put("arguments", args)
        }
        return rpc.sendRequest("tools/call", params)
    }

    suspend fun healthCheck(): Boolean {
        return try {
            rpc.sendRequest("tools/list", buildJsonObject {}, timeoutMs = 5000)
            true
        } catch (_: Exception) {
            false
        }
    }

    private fun spawnProcess() {
        val pb = ProcessBuilder(listOf(config.command) + config.args)
        pb.environment().putAll(config.env)
        pb.redirectErrorStream(false)

        val proc = pb.start()
        this.process = proc
        rpc.attach(proc.outputStream, proc.inputStream.bufferedReader(), scope)

        // Monitor stderr
        scope.launch(Dispatchers.IO) {
            proc.errorStream.bufferedReader().use { reader ->
                reader.lineSequence().forEach { line ->
                    if (line.isNotBlank()) logger.debug("[stderr] {}", line)
                }
            }
        }

        // Monitor exit
        scope.launch(Dispatchers.IO) {
            val exitCode = proc.waitFor()
            if (state == ServerState.STOPPING) return@launch
            logger.warn("Process exited with code {}", exitCode)
            handleCrash()
        }
    }

    private suspend fun initialize(): Boolean {
        return try {
            val params = buildJsonObject {
                put("protocolVersion", JsonPrimitive("2025-03-26"))
                put("capabilities", buildJsonObject {})
                put("clientInfo", buildJsonObject {
                    put("name", JsonPrimitive("mcp-bridge-kotlin"))
                    put("version", JsonPrimitive("1.0.0"))
                })
            }
            val result = rpc.sendRequest("initialize", params, config.timeout)
            rpc.sendNotification("notifications/initialized", buildJsonObject {})
            result != null
        } catch (e: Exception) {
            logger.error("Initialize error: {}", e.message)
            false
        }
    }

    private suspend fun fetchTools(): Boolean {
        return try {
            val result = rpc.sendRequest("tools/list", buildJsonObject {})
            val toolsArray = result?.jsonObject?.get("tools")?.jsonArray ?: return false
            tools = toolsArray.map { el ->
                val obj = el.jsonObject
                ToolDefinition(
                    name = obj["name"]!!.jsonPrimitive.content,
                    description = obj["description"]?.jsonPrimitive?.contentOrNull,
                    inputSchema = obj["inputSchema"]?.jsonObject,
                )
            }
            logger.info("Discovered {} tools", tools.size)
            true
        } catch (_: Exception) {
            false
        }
    }

    private fun handleCrash() {
        rpc.rejectAll("Process terminated")
        if (state == ServerState.STOPPING || state == ServerState.DEAD) return
        state = ServerState.CRASHED
        onCrashed?.invoke(name)
    }

    private suspend fun killProcess() {
        val proc = process ?: return
        this.process = null
        rpc.detach()

        proc.destroy()
        val exited = withTimeoutOrNull(5000) {
            withContext(Dispatchers.IO) { proc.waitFor() }
        }
        if (exited == null) {
            logger.warn("Force killing (destroyForcibly)")
            proc.destroyForcibly()
        }
    }
}

File: mcp-client-bridge/src/local/server-process.ts
/**
 * Single local MCP server process with state machine lifecycle.
 * Manages spawn, initialize, health check, and restart for one server.
 */

import { ChildProcess, spawn } from 'node:child_process';
import { EventEmitter } from 'node:events';
import { StdioJsonRpc } from './stdio-json-rpc.js';

export enum ServerState {
  STARTING = 'STARTING',
  READY = 'READY',
  ACTIVE = 'ACTIVE',
  CRASHED = 'CRASHED',
  RESTARTING = 'RESTARTING',
  STOPPING = 'STOPPING',
  DEAD = 'DEAD',
  FAILED = 'FAILED',
}

export interface ServerConfig {
  command: string;
  args?: string[];
  env?: Record<string, string>;
  timeout?: number;
  maxRetries?: number;
  disabled?: boolean;
}

export interface ToolDefinition {
  name: string;
  description?: string;
  inputSchema?: Record<string, unknown>;
}

export class ServerProcess extends EventEmitter {
  private process: ChildProcess | null = null;
  private rpc = new StdioJsonRpc();
  private state: ServerState = ServerState.STARTING;
  private retryCount = 0;
  private tools: ToolDefinition[] = [];

  constructor(
    readonly name: string,
    private readonly config: ServerConfig,
  ) {
    super();
  }

  get currentState(): ServerState { return this.state; }
  get currentTools(): ToolDefinition[] { return [...this.tools]; }
  get maxRetries(): number { return this.config.maxRetries ?? 3; }

  async start(): Promise<boolean> {
    this.state = ServerState.STARTING;
    this.emit('stateChange', this.name, this.state);

    try {
      this.spawnProcess();
      const initOk = await this.initialize();
      if (!initOk) {
        this.state = ServerState.FAILED;
        this.emit('stateChange', this.name, this.state);
        return false;
      }
      this.state = ServerState.READY;
      const toolsOk = await this.fetchTools();
      if (toolsOk) this.state = ServerState.ACTIVE;
      this.emit('stateChange', this.name, this.state);
      return true;
    } catch (err) {
      console.error(`[local:${this.name}] Start failed:`, err);
      this.state = ServerState.FAILED;
      this.emit('stateChange', this.name, this.state);
      return false;
    }
  }

  async stop(): Promise<void> {
    this.state = ServerState.STOPPING;
    this.emit('stateChange', this.name, this.state);
    await this.killProcess();
    this.tools = [];
  }

  async restart(): Promise<boolean> {
    if (this.retryCount >= this.maxRetries) {
      this.state = ServerState.DEAD;
      this.emit('stateChange', this.name, this.state);
      console.error(`[local:${this.name}] Max retries (${this.maxRetries}) exceeded → DEAD`);
      return false;
    }
    this.state = ServerState.RESTARTING;
    this.emit('stateChange', this.name, this.state);

    const delay = Math.min(1000 * Math.pow(2, this.retryCount), 30_000);
    this.retryCount++;
    console.error(`[local:${this.name}] Restart #${this.retryCount} in ${delay}ms`);
    await this.sleep(delay);
    return this.start();
  }

  async callTool(toolName: string, args: Record<string, unknown>): Promise<unknown> {
    return this.rpc.sendRequest('tools/call', { name: toolName, arguments: args });
  }

  async healthCheck(): Promise<boolean> {
    try {
      await this.rpc.sendRequest('tools/list', {}, 5000);
      return true;
    } catch {
      return false;
    }
  }

  private spawnProcess(): void {
    const env = { ...process.env, ...this.config.env };
    this.process = spawn(this.config.command, this.config.args ?? [], {
      stdio: ['pipe', 'pipe', 'pipe'],
      env,
      shell: process.platform === 'win32',
    });

    this.rpc.attach(this.process);
    this.process.stderr?.on('data', (chunk: Buffer) => {
      console.error(`[local:${this.name}:stderr] ${chunk.toString().trim()}`);
    });
    this.process.on('exit', (code) => this.onExit(code));
    this.process.on('error', (err) => {
      console.error(`[local:${this.name}] Process error:`, err.message);
      this.handleCrash();
    });
  }

  private async initialize(): Promise<boolean> {
    try {
      const result = await this.rpc.sendRequest('initialize', {
        protocolVersion: '2025-03-26',
        capabilities: {},
        clientInfo: { name: 'mcp-bridge-local', version: '1.0.0' },
      }, this.config.timeout ?? 30_000);
      this.rpc.sendNotification('notifications/initialized', {});
      return !!result;
    } catch (err) {
      console.error(`[local:${this.name}] Initialize timeout/error:`, err);
      return false;
    }
  }

  private async fetchTools(): Promise<boolean> {
    try {
      const result = await this.rpc.sendRequest('tools/list', {}) as { tools?: ToolDefinition[] };
      this.tools = result?.tools ?? [];
      console.error(`[local:${this.name}] Discovered ${this.tools.length} tools`);
      return true;
    } catch {
      return false;
    }
  }

  private onExit(code: number | null): void {
    if (this.state === ServerState.STOPPING) return;
    console.error(`[local:${this.name}] Process exited with code ${code}`);
    this.handleCrash();
  }

  private handleCrash(): void {
    this.rpc.rejectAll('Process terminated');
    if (this.state === ServerState.STOPPING || this.state === ServerState.DEAD) return;
    this.state = ServerState.CRASHED;
    this.emit('stateChange', this.name, this.state);
    this.emit('crashed', this.name);
  }

  private async killProcess(): Promise<void> {
    if (!this.process) return;
    const proc = this.process;
    this.process = null;
    this.rpc.detach();

    proc.kill('SIGTERM');
    const exited = await Promise.race([
      new Promise<boolean>((r) => proc.on('exit', () => r(true))),
      this.sleep(5000).then(() => false),
    ]);

    if (!exited) {
      console.error(`[local:${this.name}] Force killing (SIGKILL)`);
      proc.kill('SIGKILL');
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((r) => setTimeout(r, ms));
  }
}
